cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,421 @@
+      lcol.type(), comparator, lhs_element_index, rhs_element_index);
+  }
+
+ private:
+  column_device_view const _lhs;
+  column_device_view const _rhs;
+  Nullate const _nulls;
+  null_order const _null_precedence;
+  int const _depth;
+};
+
+/**
+ * @brief Computes whether one row is lexicographically *less* than another row.
+ *
+ * Lexicographic ordering is determined by:
+ * - Two rows are compared element by element.
+ * - The first mismatching element defines which row is lexicographically less
+ * or greater than the other.
+ *
+ * Lexicographic ordering is exactly equivalent to doing an alphabetical sort of
+ * two words, for example, `aac` would be *less* than (or precede) `abb`. The
+ * second letter in both words is the first non-equal letter, and `a < b`, thus
+ * `aac < abb`.
+ *
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <typename Nullate>
+class device_row_comparator {
+  friend class self_comparator;
+
+  /**
+   * @brief Construct a function object for performing a lexicographic
+   * comparison between the rows of two tables.
+   *
+   * @param has_nulls Indicates if either input table contains columns with nulls.
+   * @param lhs The first table
+   * @param rhs The second table (may be the same table as `lhs`)
+   * @param depth Optional, device array the same length as a row that contains starting depths of
+   * columns if they're nested, and 0 otherwise.
+   * @param column_order Optional, device array the same length as a row that indicates the desired
+   * ascending/descending order of each column in a row. If `nullopt`, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If `nullopt`, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   */
+  device_row_comparator(
+    Nullate has_nulls,
+    table_device_view lhs,
+    table_device_view rhs,
+    std::optional<device_span<int const>> depth                  = std::nullopt,
+    std::optional<device_span<order const>> column_order         = std::nullopt,
+    std::optional<device_span<null_order const>> null_precedence = std::nullopt) noexcept
+    : _lhs{lhs},
+      _rhs{rhs},
+      _nulls{has_nulls},
+      _depth{depth},
+      _column_order{column_order},
+      _null_precedence{null_precedence}
+  {
+  }
+
+ public:
+  /**
+   * @brief Checks whether the row at `lhs_index` in the `lhs` table compares
+   * lexicographically less than the row at `rhs_index` in the `rhs` table.
+   *
+   * @param lhs_index The index of row in the `lhs` table to examine
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @return `true` if row from the `lhs` table compares less than row in the `rhs` table
+   */
+  __device__ bool operator()(size_type const lhs_index, size_type const rhs_index) const noexcept
```

Comment:
Too many empty lines in this function.
Also, please use more of `auto const` for consistency.

Reply:
Made the const changes. Empty lines is subjective. For some, it's more readable.